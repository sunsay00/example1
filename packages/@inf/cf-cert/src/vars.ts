// this file has been automatically generated

export const vars = {
  CertificateArn: "arn:aws:acm:us-east-1:471589685146:certificate/4ce67418-d625-4ae2-b9b6-766b8cd4e15a",
};