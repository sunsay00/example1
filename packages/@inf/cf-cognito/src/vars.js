// this file has been automatically generated

export const vars = {
  CognitoIdentityPoolId: "us-east-1:744bb2da-6b84-4150-ba40-5f6c00d79e87",
  MobileUserPoolClientId: "5aeqscghjc8lguth4f6qosc8d7",
  UserPoolId: "us-east-1_vaHQ7ND3L",
  WebUserPoolClientId: "1djqk831709slrd9olr4q0qh7d",
};