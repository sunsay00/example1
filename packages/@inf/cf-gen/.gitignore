back
front
infra
lib
src/vars.env
src/vars.js
src/vars.ts
vars.env
vars.js
vars.ts