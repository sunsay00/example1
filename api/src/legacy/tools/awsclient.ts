import * as AWS from 'aws-sdk';
import vars from '../../_vars';
import { ICacheClient } from '@inf/cf-gen';
import * as M from '@inf/cf-gen/model';

let _sns: AWS.SNS | undefined = undefined;
let _cisp: AWS.CognitoIdentityServiceProvider;

const getSns = (region: string): AWS.SNS => {
  if (_sns == undefined) {
    _sns = new AWS.SNS({ apiVersion: '2010-03-31', region });
  }
  return _sns;
}

const getCisp = (): AWS.CognitoIdentityServiceProvider => {
  if (_cisp == undefined) {
    AWS.config.update({ region: vars.AWS_REGION });
    _cisp = new AWS.CognitoIdentityServiceProvider({ apiVersion: '2016-04-18' });
  }
  return _cisp;
};

const createEndpoint = async (region: string, PlatformApplicationArn: string, userAgent: string, deviceToken: string, customUserData: string): Promise<string> => {
  let endpointArn: undefined | string = undefined;

  let platform: 'ios' | 'android' | undefined;
  if (userAgent.toLowerCase().search('ios') != -1) {
    platform = 'ios';
  } else if (userAgent.toLowerCase().search('android') != -1) {
    platform = 'android';
  }

  if (platform == undefined) {
    throw new Error(`failed to create device endpoint - unrecognized platform for ${userAgent}`);
  }

  try {
    console.log(`Creating endpoint with token ${deviceToken} userAgent:${userAgent}`);

    const params = {
      //PlatformApplicationArn: config(platform == 'ios' ? 'AWS_SNS_APP' : 'AWS_SNS_APP_ANDROID'),
      PlatformApplicationArn,
      CustomUserData: customUserData,
      Token: deviceToken,
    };
    const result = await getSns(region).createPlatformEndpoint(params).promise();
    endpointArn = result.EndpointArn;
  } catch (err) {
    console.info(err);
    if (err.code == 'InvalidParameter') {
      console.error(`exception message ${err.message}`);
      const matches = /.*Endpoint (arn:aws:sns[^ ]+) already exists with the same Token.*/.exec(`${err.message}`);
      if (matches != undefined && matches.length >= 2) {
        endpointArn = matches[1];
      } else {
        throw err;
      }
    } else {
      throw err;
    }
  }

  if (endpointArn == undefined) {
    throw new Error('failed to retrieve endpoint arn');
  }

  return endpointArn;
};

export default class AWSClient {
  private _stage: string;
  private _region: string;
  private _platformApplicationArn: string;
  private _cache: ICacheClient | undefined;
  constructor(stage: string, region: string, platformApplicationArn: string, cache?: ICacheClient) {
    this._stage = stage;
    this._region = region;
    this._platformApplicationArn = platformApplicationArn;
    this._cache = cache;
  }
  findOrCreateTopic = async (name: string): Promise<string> => {
    const cret = this._cache && await this._cache.get(`sns.topic[${name}]`);
    if (cret != undefined) {
      return cret;
    } else {
      const ret = await getSns(this._region).createTopic({
        Name: `${this._stage}-${name}`,
      }).promise();
      if (ret.TopicArn == undefined) {
        throw new Error('failed to create topic');
      }
      this._cache && await this._cache.set(`sns.topic[${name}]`, ret.TopicArn);
      return ret.TopicArn;
    }
  }
  deleteTopic = async (topicArn: string) => {
    await getSns(this._region).deleteTopic({
      TopicArn: topicArn
    }).promise();
  }
  subscribeToTopic = async (topicArn: string, endpoint: string): Promise<string> => {
    const ret = await getSns(this._region).subscribe({
      Endpoint: endpoint,
      Protocol: 'application',
      TopicArn: topicArn,
    }).promise();
    if (ret.SubscriptionArn == undefined) {
      throw new Error('failed to subscribe to topic');
    }
    return ret.SubscriptionArn;
  }
  findSubscriptionsByTopic = async (topicArn: string, cursor?: string): Promise<{ cursor?: string, subscriptions: { endpoint: string, subscriptionArn: string }[] }> => {
    let ret = [] as { endpoint: string, subscriptionArn: string }[];
    const rslt = await getSns(this._region).listSubscriptionsByTopic({
      TopicArn: topicArn,
      NextToken: cursor,
    }).promise();
    if (rslt.Subscriptions == undefined) {
      return { subscriptions: ret };
    } else {
      for (let i = 0; i < rslt.Subscriptions.length; ++i) {
        var sub = rslt.Subscriptions[i];
        if (sub.Endpoint != undefined && sub.SubscriptionArn != undefined) {
          ret.push({
            endpoint: sub.Endpoint,
            subscriptionArn: sub.SubscriptionArn
          });
        }
      }
      return { cursor: rslt.NextToken, subscriptions: ret };
    }
  }
  unsubscribeFromTopic = async (subscriptionArn: string) => {
    await getSns(this._region).unsubscribe({
      SubscriptionArn: subscriptionArn
    }).promise();
  }
  publish = async (topicArn: string, message: string, custom: unknown): Promise<string> => {
    const apns = JSON.stringify({
      ...custom,
      aps: { alert: message, sound: 'default', badge: 0 },
    });

    const isProd = this._stage == 'production';

    const payload = {
      ...(isProd ? { APNS: apns } : { APNS_SANDBOX: apns }),
      GCM: JSON.stringify({ data: { ...custom, message } }),
      default: message,
    };

    const ret = await getSns(this._region).publish({
      MessageStructure: 'json',
      Message: JSON.stringify(payload),
      TopicArn: topicArn,
    }).promise();
    if (ret.MessageId == undefined)
      throw new Error('bad publish message response');
    return ret.MessageId;
  }

  send = async (targetArn: string, message: string, custom: unknown) => {
    const apns = JSON.stringify({
      ...custom,
      aps: { alert: message, sound: 'default', badge: 0 },
    });

    const isProd = this._stage == 'production';

    const payload = {
      ...(isProd ? { APNS: apns } : { APNS_SANDBOX: apns }),
      GCM: JSON.stringify({ data: { ...custom, message } }),
      default: message,
    };

    const ret = await getSns(this._region).publish({
      MessageStructure: 'json',
      Message: JSON.stringify(payload),
      TargetArn: targetArn,
    }).promise();
    if (ret.MessageId == undefined) {
      throw new Error('bad send message response');
    }
    return ret.MessageId;
  }
  unregisterDevice = async (endpoint: string): Promise<void> => {
    await getSns(this._region).deleteEndpoint({
      EndpointArn: endpoint,
    }).promise();
  }
  registerDevice = async (userAgent: string, endpoint: string | undefined, deviceToken: string, customUserData: string): Promise<string> => {
    let endpointArn = endpoint;

    let updateNeeded = false;
    let createNeeded = true;

    if (endpointArn == undefined) {
      // No endpoint ARN is stored; need to call CreateEndpoint
      endpointArn = await createEndpoint(this._region, this._platformApplicationArn, userAgent, deviceToken, customUserData);
      createNeeded = false;
    }

    console.log("Retrieving endpoint data...");
    // Look up the endpoint and make sure the data in it is current, even if
    // it was just created

    try {
      const data: any = await getSns(this._region).getEndpointAttributes({
        EndpointArn: endpointArn
      }).promise();
      updateNeeded = data.Token == deviceToken || !data.Enabled;
    } catch (err) {
      console.info(err);
      if (err.code == 'NotFound') {
        createNeeded = true;
      } else {
        throw err;
      }
    }

    if (createNeeded) {
      endpointArn = await createEndpoint(this._region, this._platformApplicationArn, userAgent, deviceToken, customUserData);
    }

    console.log("updateNeeded=" + updateNeeded);

    if (updateNeeded) {
      // endpoint is out of sync with the current data;
      // update the token and enable it.
      console.log(`Updating endpoint ${endpointArn}`);
      const params: any = {
        Attribures: {
          Token: deviceToken,
          Enabled: true,
        },
        EndpointArn: endpointArn,
      };
      await getSns(this._region).setEndpointAttributes(params);
    }

    return endpointArn;
  }
};

const padLeft = (str: string, len: number, fill: string = ' ') =>
  str.length >= len ? str : (new Array(Math.ceil((len - str.length) / fill.length) + 1).join(fill)).substr(0, (len - str.length)) + str;

export const inviteUser = async (username: string, email: string, locale: string, role: string, uservar1: string, action: 'RESEND' | 'SUPRESS' | undefined): Promise<M.PendingSeller | undefined> => {
  const cisp = getCisp();
  const poolId = vars.UserPoolId;

  // adminCreateUser throws UsernameExistsException even when the user does not exist,
  // as a workaround we check if the user doesn't exist with adminGetUser and ignore the exception

  /*
  var found = false;
  try {
    const user = await cisp.adminGetUser({
      Username: username,
      UserPoolId: poolId,
    }).promise();
    console.log('::::', JSON.stringify(user));
    found = true;
  } catch (err) {
    console.log('BBBB', JSON.stringify(err));
    // ignoring exception
  }

  if (found) {
    throw new Error('username unavailable');
  }
  */

  try {
    const splitedName = username.split(' ');
    let login = splitedName.length > 1 ? splitedName[0] + '.' + splitedName[splitedName.length - 1] : username;
    if (action == 'RESEND') {
      login = username;
    } else {
      const listUsersResponse = await cisp.listUsers({
        UserPoolId: poolId,
        Filter: 'username ^= "' + login + '"',
      }).promise();
      var regex = new RegExp(login + '+' + '[1-9]', 'i');
      const found = listUsersResponse.Users && listUsersResponse.Users.filter(u => u.Username
        && (u.Username.match(regex)
          || u.Username == login));

      if (found != undefined && found.length > 0) {
        login = login + found.length.toString();
      }
      console.log('login to be created: ' + login);
    }
    const response = await cisp.adminCreateUser({
      DesiredDeliveryMediums: ['EMAIL'],
      ForceAliasCreation: true,
      MessageAction: action,
      TemporaryPassword: padLeft(`${Math.ceil(Math.random() * 1000000001) + 1}`, 8, '0'),
      UserAttributes: [
        { Name: 'email', Value: email },
        { Name: 'email_verified', Value: 'True' },
        { Name: 'locale', Value: locale },
        { Name: 'family_name', Value: uservar1 },
        { Name: 'preferred_username', Value: login },
        { Name: 'custom:role', Value: role },
        { Name: 'custom:uservar1', Value: uservar1 },
        { Name: 'custom:uservar2', Value: username },
      ],
      Username: login,
      UserPoolId: poolId,
    }).promise();
    if (action == 'RESEND') {
      console.log('temp password resend for ' + login);
    } else {
      console.log('created ' + login);
    }
    if (response.User != undefined) {
      const sub = response.User.Attributes && response.User.Attributes.find(a => a.Name == 'sub');
      const email = response.User.Attributes && response.User.Attributes.find(a => a.Name == 'email');
      const name = response.User.Attributes && response.User.Attributes.find(a => a.Name == 'custom:uservar2');
      const username = response.User.Username;
      if (sub && sub.Value && email && email.Value && name && name.Value && username) {
        return {
          sub: sub.Value,
          username: username,
          email: email.Value,
          name: name.Value
        }
      } else {
        return undefined;
      }
    } else {
      return undefined;
    }
  } catch (err) {
    console.log(`error creating user ${username}`, err);
    throw err;
  }
};

export const adminDeleteUser = async (username: string) => {
  const cisp = getCisp();
  await cisp.adminDeleteUser({
    UserPoolId: vars.UserPoolId,
    Username: username,
  }).promise();
}

export const adminEnableDisableUser = async (sub: string, dealershipId: string, disableUser: boolean) => {
  const users = await adminListUsers('CONFIRMED', dealershipId);
  const found = users.find(u => u.sub == sub);
  if (found != undefined) {
    const cisp = getCisp();
    disableUser
      ? await cisp.adminDisableUser({
        UserPoolId: vars.UserPoolId,
        Username: found.username != undefined ? found.username : '',
      }).promise()
      : await cisp.adminEnableUser({
        UserPoolId: vars.UserPoolId,
        Username: found.username != undefined ? found.username : '',
      }).promise();
  } else {
    console.log('Cannot find user for ' + sub);
    throw new Error('Invalid user');
  }
}

export const adminConfirmSignUp = async (username: string) => {
  const cisp = getCisp();
  await cisp.adminConfirmSignUp({
    UserPoolId: vars.UserPoolId,
    Username: username,
  }).promise();
}

export const adminAddUserToGroup = async (username: string, group: string) => {
  const cisp = getCisp();
  await cisp.adminAddUserToGroup({
    UserPoolId: vars.UserPoolId,
    Username: username,
    GroupName: group,
  }).promise();
}

export const adminListUsers = async (filter: string, dealershipId: string, limit?: number, token?: string): Promise<{ sub: string, username: string, email: string, name: string, token?: string }[]> => {
  const cisp = getCisp();
  const listUsersResponse = await cisp.listUsers({
    UserPoolId: vars.UserPoolId,
    Filter: 'family_name = "' + dealershipId + '" and cognito:user_status = "' + filter + '"',
    Limit: limit,
    PaginationToken: token
  }).promise();
  if (listUsersResponse.Users != undefined) {
    const filteredUsers = listUsersResponse.Users.filter(u => u.Attributes
      && u.Attributes.find(a => a.Name == 'custom:uservar1'
        && a.Value != undefined && a.Value == dealershipId) != undefined);
    if (filteredUsers.length > 0) {
      let users: { sub: string, username: string, email: string, name: string, token?: string }[];
      users = [];
      filteredUsers.map((u) => {
        const username = u.Username;
        const sub = u.Attributes && u.Attributes.find(a => a.Name == 'sub');
        const email = u.Attributes && u.Attributes.find(a => a.Name == 'email');
        const name = u.Attributes && u.Attributes.find(a => a.Name == 'custom:uservar2');
        if (username != undefined && sub != undefined && sub.Value != undefined && email != undefined && email.Value != undefined) {
          users.push({
            sub: sub.Value,
            username: username,
            email: email.Value,
            name: name != undefined && name.Value != undefined ? name.Value : username,
            token: listUsersResponse.PaginationToken
          });
        }
      });
      return users;
    } else {
      return [];
    }
  } else {
    return [];
  }
}
