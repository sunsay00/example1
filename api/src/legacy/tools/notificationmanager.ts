import AWSClient from './awsclient';
import { INotificationManager, IStore, ICacheClient } from '@inf/cf-gen';
import * as M from '@inf/cf-gen/model';
import { UserContext } from '../types';

export default class NotificationManager implements INotificationManager<UserContext> {
  private _aws: AWSClient;
  private _cache: ICacheClient | undefined;
  constructor(params: { stage: string, region: string, platformApplicationArn: string, cache?: ICacheClient }) {
    this._aws = new AWSClient(params.stage, params.region, params.platformApplicationArn, params.cache);
    this._cache = params.cache;
  }

  async dispatchDeckFactories(store: IStore<UserContext>, sub: string, action: string, obj: M.DeckFactory) { }
  async dispatchDecks(store: IStore<UserContext>, sub: string, action: string, obj: M.Deck) { }
  async dispatchCommentLists(store: IStore<UserContext>, sub: string, action: string, obj: M.CommentList) { }
  async dispatchDealerships(store: IStore<UserContext>, sub: string, action: string, obj: M.Dealership) { }
  async dispatchCarInventories(store: IStore<UserContext>, sub: string, action: string, obj: M.CarInventory) { }
  async dispatchSellers(store: IStore<UserContext>, sub: string, action: string, obj: M.Seller) { }
  async dispatchNotes(store: IStore<UserContext>, sub: string, action: string, obj: M.Note) { }
  async dispatchCarInfos(store: IStore<UserContext>, sub: string, action: string, obj: M.CarInfo) { }

  async subscribe($user: UserContext, store: IStore<UserContext>, chatIdentity: string, typename: string, id: string) {
    const name = `${typename}-${id}`;
    const topicArn = await this._aws.findOrCreateTopic(name);
    const devicelist = await store.deviceListsFindByChatIdentity($user, chatIdentity);
    if (devicelist != undefined) {
      for (let key in devicelist.devices) {
        const device = devicelist.devices[key];
        await this._aws.subscribeToTopic(topicArn, device.endpoint);
      }
    }
  }

  async unsubscribe($user: UserContext, store: IStore<UserContext>, chatIdentity: string, typename: string, id: string) {
    const name = `${typename}-${id}`;
    //const topicArn = 
    await this._aws.findOrCreateTopic(name);
    const devicelist = await store.deviceListsFindByChatIdentity($user, chatIdentity);
    if (devicelist != undefined) {
      let devices = [...devicelist.devices];
      let found = await this._aws.findSubscriptionsByTopic(name);
      while (devices.length > 0) {
        for (let key in found.subscriptions) {
          const sub = found.subscriptions[key];
          if (devices.findIndex(d => d.endpoint == sub.endpoint) != -1) {
            await this._aws.unsubscribeFromTopic(sub.subscriptionArn);
            devices = devices.filter(d => d.endpoint != sub.endpoint);
            if (devices.length == 0) {
              break;
            }
          }
        }
        if (found.cursor != undefined) {
          found = await this._aws.findSubscriptionsByTopic(name, found.cursor);
        } else {
          break;
        }
      }
    }
  }

  /*async publishToBuilding(buildingId: string, type: string, id: string, message: string) {
    const name = `buildings-${buildingId}`;
    const topicArn = await this._aws.findOrCreateTopic(name);
    const custom = { type, id, parentType: 'buildings', parentId: buildingId };
    await this._aws.publish(topicArn, message, custom);
  }*/

  async sendToUser($user: UserContext, store: IStore<UserContext>, chatIdentity: string, type: string, id: string, message: string) {
    const list = await store.deviceListsFindByChatIdentity($user, chatIdentity);
    if (list != undefined) {
      const custom = { type, id, parentType: '', parentId: '' };
      for (let key in list.devices) {
        const device = list.devices[key];
        await this._aws.send(device.endpoint, message, custom);
      }
    }
  }

  registerDevice = async (userAgent: string, chatIdentity: string, endpoint: string | undefined, deviceToken: string): Promise<string> => {
    if (endpoint == undefined) {
      const retEndpoint = await this._aws.registerDevice(userAgent, endpoint, deviceToken, chatIdentity);
      this._cache && await this._cache.set(`endpoints.chatIdentity[${chatIdentity}]`, retEndpoint);
      return retEndpoint;
    } else {
      const prevEndpoint = this._cache && await this._cache.get(`endpoints.chatIdentity[${chatIdentity}]`);
      if (prevEndpoint == endpoint) {
        return prevEndpoint;
      } else {
        const retEndpoint = await this._aws.registerDevice(userAgent, endpoint, deviceToken, chatIdentity);
        this._cache && await this._cache.set(`endpoints.chatIdentity[${chatIdentity}]`, retEndpoint);
        return retEndpoint;
      }
    }
  }
  unregisterDevice = async (chatIdentity: string, endpoint: string): Promise<void> => {
    await this._aws.unregisterDevice(endpoint);
    this._cache && await this._cache.del([`endpoints.chatIdentity[${chatIdentity}]`]);
  }
}
