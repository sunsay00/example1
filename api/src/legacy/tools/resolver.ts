import { createResolver } from '@inf/cf-gen';
import NotificationManager from './notificationmanager';
import UsersService from '../services/users';
import SellersService from '../services/sellers';
import SystemService from '../services/system';
import DeviceListsService from '../services/devicelists';

import { ICacheClient, IDBClient } from '@inf/cf-gen';

export const createDefaultResolver = (params: {
  stage: string,
  region: string,
  locale: string,
  platformApplicationArn: string,
  db: IDBClient,
  cache?: ICacheClient,
  onInit?: () => Promise<void>,
}) => {
  const notifications = new NotificationManager(
    params.stage,
    params.region,
    params.platformApplicationArn,
    params.cache);
  return createResolver({
    stage: params.stage,
    notifications,
    db: params.db,
    onInit: params.onInit,
    services: {
      users: store => new UsersService(notifications, store, params.locale, params.region),
      systems: store => new SystemService(notifications, store),
      deviceLists: store => new DeviceListsService(notifications, store),
      sellers: store => new SellersService(notifications, store),
    },
  });
};
