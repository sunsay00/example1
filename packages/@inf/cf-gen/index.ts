import RDSStore from './back/api/src/api/stores/rdsstore';
import CacheStore from './back/api/src/api/stores/cachestore';
import { IUserContext, INotificationManager, IDBClient, ICacheClient } from './types';
import { createServiceMapper, MappedServices } from './back/api/src/api/mapper';
import { Resolver } from './tools/resolver';
import * as fix from './back/api/src/api/__tests__/fixtures';

export * from './types';
export * from './back/api/src/types/serviceinterfaces';
export const fixtures = fix;

export const createCachedResolver = <C extends IUserContext>(
  stage: string,
  notifications: INotificationManager<C>,
  cache: ICacheClient,
  db: IDBClient,
  services: MappedServices<C>,
  onPreResolve?: () => Promise<void>,
  onPostResolve?: () => Promise<void>,
) => {
  const store = new RDSStore(stage, db);
  return new Resolver(
    stage,
    store,
    createServiceMapper(notifications, new CacheStore(store, cache), services),
    onPreResolve,
    onPostResolve);
};

export const createResolver = <C extends IUserContext>(params: {
  stage: string,
  notifications: INotificationManager<C>,
  db: IDBClient,
  services: MappedServices<C>,
  onPreResolve?: () => Promise<void>,
  onPostResolve?: () => Promise<void>,
}) => {
  const store = new RDSStore(params.stage, params.db);
  return new Resolver(
    params.stage,
    store,
    createServiceMapper(params.notifications, store, params.services),
    params.onPreResolve,
    params.onPostResolve);
}