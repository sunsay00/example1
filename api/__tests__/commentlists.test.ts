import "jest";
import { verifyVars } from '@inf/common';
import { createDefaultResolver } from '../src/legacy/tools/resolver';
import RDSDBClient from '@inf/cf-serverless-postgres';
import { fixtures as fix, IUserContext, User, CommentItemInput } from '@inf/cf-gen';
import vars from '../src/_vars';

const config = verifyVars({
  STAGE: process.env.STAGE,
  AWS_REGION: process.env.AWS_REGION
});

const db = new RDSDBClient(vars.DB_TEST_URL);
const cache = undefined; // new CacheClient(config.redisUrl),

const resolver = createDefaultResolver({
  stage: config.STAGE,
  region: config.AWS_REGION,
  locale: 'en',
  platformApplicationArn: '',
  db,
  cache,
});

describe('commentlists', () => {
  let userCtx: IUserContext;
  let user: User;
  beforeAll(async () => {
    await db.init();
    userCtx = { sub: 'commenter', groups: [] };
    user = await fix.backendCreateUser(resolver.store(), userCtx, "the user", "avataruri.jpg", "en", "us-east-1");
  });

  afterAll(async () => {
    await fix.backendRemoveUser(resolver.store(), userCtx, user.id);
    await db.deinit();
  });

  it('should CRUD a comment list', async () => {
    let comments: CommentItemInput[] = [{ text: 'mytext1' }, { text: 'mytext1' }];
    const { id, ...rest } = await fix.commentListsCreate(resolver, userCtx, 'myparentid', comments);
    const commentList = {
      ...rest, comments: rest.comments.map(c => {
        const { createdAt, updatedAt, ...rest } = c;
        return rest;
      })
    };
    expect(commentList).toMatchSnapshot();
    await fix.commentListsRemove(resolver, userCtx, id);
  });

  it('should return undefined if a comment is added to a non-existant list', async () => {
    const comment: CommentItemInput = {
      text: 'mytext1',
    };
    const appendedComment = await fix.commentListsAppendComment(resolver, userCtx, "5896757689", comment);
    expect(appendedComment).toBeUndefined();
  });

  it('should append a comment to an existing list', async () => {
    const comment: CommentItemInput = {
      text: 'mytext11',
    };
    const commentList = await fix.commentListsCreate(resolver, userCtx, 'myparentid2', [comment]);
    expect(commentList.comments).toHaveLength(1);
    const anotherComment: CommentItemInput = {
      text: 'mytext22',
    };
    const appendedComment = await fix.commentListsAppendComment(resolver, userCtx, commentList.id, anotherComment);
    if (appendedComment == undefined) throw new Error('appended comment not found');
    const { createdAt, updatedAt, ...appendedCommentRest } = appendedComment;
    expect(appendedCommentRest).toMatchSnapshot();
    const foundCommentList = await fix.commentListsFindById(resolver, userCtx, commentList.id);
    expect(foundCommentList).toBeDefined();
    if (foundCommentList != undefined) {
      expect(foundCommentList.comments).toHaveLength(2);
    }

    anotherComment.text = 'mytext33';
    const updatedComment = await fix.commentListsUpdateComment(resolver, userCtx, commentList.id, 1, anotherComment);
    if (updatedComment != undefined) {
      const { createdAt: _1, updatedAt: _2, ...updatedCommentRest } = updatedComment;
      expect(updatedCommentRest).toMatchSnapshot();
    }
    const foundCommentList2 = await fix.commentListsFindById(resolver, userCtx, commentList.id);
    expect(foundCommentList2).toBeDefined();
    if (foundCommentList2 != undefined) {
      expect(foundCommentList2.comments).toHaveLength(2);
      expect(foundCommentList2.comments[0].text).toEqual(comment.text);
      expect(foundCommentList2.comments[1].text).toEqual(anotherComment.text);
    }

    expect(await fix.commentListsRemove(resolver, userCtx, commentList.id)).toBeDefined();
  });

});

