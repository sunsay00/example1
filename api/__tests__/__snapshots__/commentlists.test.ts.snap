// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commentlists should CRUD a comment list 1`] = `
Object {
  "comments": Array [
    Object {
      "id": "0",
      "likeCount": 0,
      "removed": false,
      "sub": "commenter",
      "text": "mytext1",
      "user": Object {
        "avatarUri": "avataruri.jpg",
        "name": "the user",
      },
    },
    Object {
      "id": "1",
      "likeCount": 0,
      "removed": false,
      "sub": "commenter",
      "text": "mytext1",
      "user": Object {
        "avatarUri": "avataruri.jpg",
        "name": "the user",
      },
    },
  ],
  "parentId": "myparentid",
}
`;

exports[`commentlists should append a comment to an existing list 1`] = `
Object {
  "id": "1",
  "likeCount": 0,
  "removed": false,
  "sub": "commenter",
  "text": "mytext22",
  "user": Object {
    "avatarUri": "avataruri.jpg",
    "name": "the user",
  },
}
`;

exports[`commentlists should append a comment to an existing list 2`] = `
Object {
  "id": "1",
  "likeCount": 0,
  "removed": false,
  "sub": "commenter",
  "text": "mytext33",
  "user": Object {
    "avatarUri": "avataruri.jpg",
    "name": "the user",
  },
}
`;
